src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs

```diff
@@ -0,0 +1,101 @@
+                DiagnosticHelpers.DefaultDiagnosticSeverity,
+                isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+                description: s_Description,
+                helpLinkUri: null,
+                customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            // Security analyzer - analyze and report diagnostics on generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(compilationStartAnalysisContext =>
+            {
+                var compilation = compilationStartAnalysisContext.Compilation;
+                var pageTypeSymbol = compilation.GetTypeByMetadataName(WellKnownTypeNames.SystemWebUIPage);
+
+                if (pageTypeSymbol == null)
+                {
+                    return;
+                }
+
+                var eventArgsTypeSymbol = compilation.GetTypeByMetadataName(WellKnownTypeNames.SystemEventArgs);
+
+                if (eventArgsTypeSymbol == null)
+                {
+                    return;
+                }
+
+                compilationStartAnalysisContext.RegisterSymbolAction(symbolAnalysisContext =>
+                {
+                    var classSymbol = (INamedTypeSymbol)symbolAnalysisContext.Symbol;
+                    var baseClassSymbol = classSymbol.BaseType;
+
+                    if (baseClassSymbol.Equals(pageTypeSymbol))
+                    {
+                        var onInitMethodSymbol = classSymbol.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(s => s.Name == "OnInit" &&
+                                                                                                        s.Parameters.Length == 1 &&
+                                                                                                        s.Parameters[0].Type.Equals(eventArgsTypeSymbol) &&
+                                                                                                        s.IsProtected());
+
+                        if (onInitMethodSymbol != null)
+                        {
+                            if (onInitMethodSymbol.GetTopmostOperationBlock(compilation)
+                                                    .Descendants()
+                                                    .Where(s => s is ISimpleAssignmentOperation simpleAssignmentOperation &&
+                                                                simpleAssignmentOperation.Target is IPropertyReferenceOperation propertyReferenceOperation &&
+                                                                propertyReferenceOperation.Property.Name == "ViewStateUserKey")
```

Comment:
I would recommend additional checks to verify the type of the property matches your expected property type and the `IPropertyReferenceOperation.Instance` is non-null `IInstanceReferenceOperation` with `ReferenceKind` [InstanceReferenceKind.ContainingTypeInstance](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/InstanceReferenceKind.cs,14)